platform/eureka/conversion.go

```diff
@@ -0,0 +1,165 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package eureka
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/golang/glog"
+	"istio.io/pilot/model"
+)
+
+// Convert Eureka applications to services. If provided, only convert applications in the hostnames whitelist,
+// otherwise convert all.
+func convertServices(apps []*application, hostnames map[string]bool) map[string]*model.Service {
+	services := make(map[string]*model.Service)
+	for _, app := range apps {
+		for _, instance := range app.Instances {
+			if len(hostnames) > 0 && !hostnames[instance.Hostname] {
+				continue
+			}
+
+			if instance.Status != statusUp {
+				continue
+			}
+
+			service := services[instance.Hostname]
+			if service == nil {
+				service = &model.Service{
+					Hostname:     instance.Hostname,
+					Address:      "",
+					Ports:        make(model.PortList, 0),
+					ExternalName: "",
+				}
+				services[instance.Hostname] = service
+			}
+
+			protocol := convertProtocol(instance.Metadata)
+			for _, port := range convertPorts(instance) {
+				if servicePort, exists := service.Ports.GetByPort(port.Port); exists {
+					if servicePort.Protocol != protocol {
+						glog.Warningf(
+							"invalid Eureka config: "+
+								"%s:%d has conflicting protocol definitions %s, %s",
+							instance.Hostname, servicePort.Port,
+							servicePort.Protocol, protocol)
+					}
+					continue
+				}
+
+				service.Ports = append(service.Ports, port)
+			}
+		}
+	}
+	return services
+}
+
+// Convert Eureka applications to service instances. The services argument must contain a map of hostnames to
+// services. Only service instances with a corresponding service are converted.
+func convertServiceInstances(services map[string]*model.Service, apps []*application) []*model.ServiceInstance {
+	out := make([]*model.ServiceInstance, 0)
+	for _, app := range apps {
+		for _, instance := range app.Instances {
+			if services[instance.Hostname] == nil {
+				continue
+			}
+
+			if instance.Status != statusUp {
+				continue
+			}
+
+			for _, port := range convertPorts(instance) {
+				out = append(out, &model.ServiceInstance{
+					Endpoint: model.NetworkEndpoint{
+						Address:     instance.IPAddress,
+						Port:        port.Port,
+						ServicePort: port,
+					},
+					Service: services[instance.Hostname],
+					Tags:    convertTags(instance.Metadata),
+				})
+			}
+		}
+	}
+	return out
+}
+
+func convertPorts(instance *instance) model.PortList {
+	out := make(model.PortList, 0, 2)
+	protocol := convertProtocol(instance.Metadata)
+	for _, port := range []*port{instance.Port, instance.SecurePort} {
+		if port == nil || !port.Enabled {
+			continue
+		}
+
+		out = append(out, &model.Port{
+			Name:     fmt.Sprint(port.Port),
+			Port:     port.Port,
+			Protocol: protocol,
+		})
+	}
+	return out
+}
+
+const protocolMetadata = "istio.protocol" // metadata key for port protocol
```

Comment:
I understand that Eureka does not support protocols. But this would require changes to existing eureka clients or require people to update instance info out of band.

As part of the effort to read Eureka configuration from a file (e.g., the server's IPs, API base URL, etc., one can also imagine that the user provides a list of ports and their protocols - which can be used during runtime to lookup the protocols associated with a port). I leave it up to you to do this in this PR or in a follow up PR to add the ability to read this info from a statically supplied config file.